proto/docs/p4runtime.md

```diff
@@ -319,3 +319,123 @@ should take into account the top-level message and act as follows:
   switch only returns the counter state associated with the match given in
   `DirectCounterEntry.TableEntry.match`, and does not populate other fields
   within the `TableEntry` (such as `Action`).
+
+## Error reporting
+
+When the p4.Write() fails for some reason, P4 Runtime returns `grpc::Status` to
+report errors. `grpc::Status` has top-level canonical `error_code` that
+represents a RPC-wide error, along with `error_details` which is serialized
+`google.rpc.Status` that contains per-entity level errors. `google.rpc.Status`
+can have multiple `p4.Error` in repeated `details` field, each repesent
+per-entity errors in the p4.Write() batch.
+
+`p4.Error` has gRPC canonical error space for P4 Runtime error reporting.
+Additionally, `p4.Error` also allow different vendors / chipmakers to express
+their own error code and error space as well.
+
+![P4 Runtime Error Proto](p4-runtime-error-proto.svg)
+
+P4 Runtime will populate the top-level error code as follows:
+
+-   Use any canonical error code that best describes RPC-wide error. If the
+    entire batch has failed due to P4 Runtime connectivity issue, P4 Runtime
+    should return a canonical code that best describes the reason, e.g.
+    `NOT_FOUND`.
+
+-   Use `UNKNOWN` top level error code for all Write RPC failures, whether
+    partial batch or totalfailures, e.g. one `RESOURCE_EXHAUSTED` and one
+    `INVALID_ARGUMENT`.
+
+-   2nd level canonical error (google.rpc.Status) should be same as top-level
+    canonical error code.
+
+When reporting per-entity level errors, P4 Runtime includes all batch request
+responses even if only a part of batch requests failed.
+
+```
+# Example of error message:
+
+{
+  error_code: 2  # UNKNOWN
+  error_message: "Write failure."
+  error_details {
+    code: 2  # UNKNOWN
+    message: "Write failure."
+    details {
+      canonical_code: 8  # RESOURCE_EXHAUSTED
+      message: "Table is full."
+      space: "foo-psa-vendor1"
```

Comment:
I would use `targetX-psa-vendorY` instead
In any case, use `target` instead of `foo`.